app/services/postcode_validator.rb

```diff
@@ -0,0 +1,9 @@
+class PostcodeValidator
+  VALID_POSTCODE = /^([A-Z]{1,2}[0-9][A-Z0-9]? ?[0-9][A-Z]{2})$/i.freeze
```

Comment:
Is this regex based on the [vulnerable people form regex](https://github.com/alphagov/govuk-coronavirus-vulnerable-people-form/blob/b4d257b89940fe12bfbc84c4b3edffcc16a0146b/app/helpers/field_validation_helper.rb#L108) (without the wider range of postcodes: overseas territories etc), or is it hand-written based on the following description in the Trello card?

> Anything that doesn't look like a correct postcode (valid options are A1 1AA, A11 1AA, AA1 1AA, AA11 1AA, A1A 1AA, AA1A 1AA, where A is any letter and 1 is any number) -- we should ignore case, spaces and other symbols, as per the [design system](https://design-system.service.gov.uk/patterns/addresses/#allow-different-postcode-formats)

I'm just wondering how much investigation has been done into valid postcode formats, and worry that we might be rejecting valid postcodes if we're not careful. The [standard StackOverflow answer](https://stackoverflow.com/questions/164979/regex-for-matching-uk-postcodes) on this one contains quite a complex regex, which suggests to me that there may be some postcodes our `VALID_POSTCODE` regex doesn't catch. I might catch up with you & Bruce about this on Slack.

Reply:
Yes, VALID_POSTCODE regex is based on the one in vulnerable people form, but excluding overseas territories etc (all bolded sections here refer to them):
^(([A-Z]{1,2}\d[A-Z\d]?**|ASCN|STHL|TDCU|BBND|[BFS]IQQ|PCRN|TKCA)**  |d[A-Z]{2}. **| BFPO ?\d{1,4}|(KY\d|MSR|VG|AI)[ -]?\d{4}|[A-Z]{2} ?\d{2}|GE ?CX|GIR ?0A{2}|SAN ?TA1)**$

I have been looking into other options (including those on StackOverflow - and it looks like more complex ones, beside including mentioned territories, ensure special characters are permitted - here I am trying to handle it in line 6 `postcode = postcode.gsub(/[~,.<>;':"\/\[\]|{}()=_+-]| /, "")`) to keep regex more readable.
Hope it clarifies and I am happy to discuss further!